mathtext: Finetuning sup/super block to match TeX reference

zblz opened the pull request:
As mentioned in #4872, I found that sub/super blocks in mathtext have slightly too much space after them, even after @mdboom fix in #4401. As initial reference, this a TeX rendering of a bunch of sub/superscripts (with E and 5 because they are box-like and separation differences are easier to spot):
![subsup_test_tex](https://cloud.githubusercontent.com/assets/576258/9117283/05c01816-3c61-11e5-821a-484d800f4378.png)

The current master renders it as follows:
![subsup_test_mpl_master](https://cloud.githubusercontent.com/assets/576258/9117307/209ab696-3c61-11e5-8317-91bbc0f47877.png)

And with the modifications of this PR, it looks like this:
![subsup_test_mpl](https://cloud.githubusercontent.com/assets/576258/9117316/2b5a300c-3c61-11e5-85d0-3a157af8d0c7.png)

Summary of changes:

* Additional kerning of -0.1 em after sub/super blocks.
* Superscripts are shifted away from nucleus by 0.15 em when nucleus is slanted.
* When there is both sub and superscript and nucleus is slanted, superscripts are only shifted an aditional 0.15 em (down from >0.20 em)
* Subscripts are shifted down 0.25 x-height instead of 0.5 when there is no superscript.

Regression images coming soon!

WeatherGod commented:
Comparing between the Tex and mathtext, It looks like the superscript is
slightly too high as well.

On Thu, Aug 6, 2015 at 12:44 PM, Victor Zabalza <notifications@github.com>
wrote:

> As mentioned in #4872 <https://github.com/matplotlib/matplotlib/pull/4872>,
> I found that sub/super blocks in mathtext have slightly too much space
> after them, even after @mdboom <https://github.com/mdboom> fix in #4401
> <https://github.com/matplotlib/matplotlib/pull/4401>. As initial
> reference, this a TeX rendering of a bunch of sub/superscripts (with E and
> 5 because they are box-like and separation differences are easier to spot):
> [image: subsup_test_tex]
> <https://cloud.githubusercontent.com/assets/576258/9117283/05c01816-3c61-11e5-821a-484d800f4378.png>
>
> The current master renders it as follows:
> [image: subsup_test_mpl_master]
> <https://cloud.githubusercontent.com/assets/576258/9117307/209ab696-3c61-11e5-8317-91bbc0f47877.png>
>
> And with the modifications of this PR, it looks like this:
> [image: subsup_test_mpl]
> <https://cloud.githubusercontent.com/assets/576258/9117316/2b5a300c-3c61-11e5-85d0-3a157af8d0c7.png>
>
> Summary of changes:
>
>    - Additional kerning of -0.1 em after sub/super blocks.
>    - Superscripts are shifted away from nucleus by 0.15 em when nucleus
>    is slanted.
>    - When there is both sub and superscript and nucleus is slanted,
>    superscripts are only shifted an aditional 0.15 em (down from >0.20 em)
>    - Subscripts are shifted down 0.25 x-height instead of 0.5 when there
>    is no superscript.
>
> Regression images coming soon!
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/matplotlib/matplotlib/pull/4873
> Commit Summary
>
>    - first attempt
>    - finetuning with cm TeX
>
> File Changes
>
>    - *M* lib/matplotlib/mathtext.py
>    <https://github.com/matplotlib/matplotlib/pull/4873/files#diff-0> (21)
>
> Patch Links:
>
>    - https://github.com/matplotlib/matplotlib/pull/4873.patch
>    - https://github.com/matplotlib/matplotlib/pull/4873.diff
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/matplotlib/matplotlib/pull/4873>.
>

zblz commented:
Well, it seems that reducing the space after the sub/subscript is not that simple after all: it works well with numbers, but for slanted characters:
![subsup_test_mpl_finetune](https://cloud.githubusercontent.com/assets/576258/9118318/796dbd44-3c67-11e5-9aa5-dbb4a847141f.png)

TeX, of course, handles these beautifully. I'll try to see where the difference in spacing comes from.

@WeatherGod: indeed it is, I'll try to match that as well.

zblz commented:
Trying to see why ``$a_{ji}$`` was being rendered with too little kerning, I found that the current master renders ``$a_{i}$`` and ``$a_i$`` differently. After much blood and tears in the depths of ``mathtext.py``, I think it is because calling ``.shrink()`` on a ``Char`` object results in different kerning that calling it on a ``Hlist`` object. The groups between ``{ }`` are parsed as ``Hlist`` objects, whereas the characters in ``$a_i$`` are parsed as ``Char``. The current master shrinks and then converts into a Hlist, and I have changed it so that evereything is an ``Hlist`` before shrinking, so there is better consistency. To showcase this, I made a plot with two lines that only differ by (unnecesary) braces:
```
r'$x_kx^py^{p-2} \quad d_i^jb_jc_kd \quad x^j_i E^0 E^0_u$',
r'${x}_{k}{x}^{p}{y}^{p-2} \quad {d}_{i}^{j}{b}_{j}{c}_{k}{d} \quad {x}^{j}_{i}{E}^{0}{E}^0_u$',
```
Current master renders them clearly different:
![subsup_test_mpl_master](https://cloud.githubusercontent.com/assets/576258/9124620/6eadcc9a-3c90-11e5-826f-bae3b7e28267.png)

And with the changes mentioned they are consistent with each other:
![subsup_test_mpl_finetune](https://cloud.githubusercontent.com/assets/576258/9124624/798c3052-3c90-11e5-9c68-93536729c22c.png)

I have also finetuned the supsuper spacings to match Computer Modern on TeX, but I have found that these do not work properly for other fontsets.... Here is the reference TeX rendering (with text.usetex=True):
![subsup_test_mpl_tex](https://cloud.githubusercontent.com/assets/576258/9124742/c73bb952-3c91-11e5-8928-48594e094d4a.png)

And with this PR:
![subsup_test_mpl_finetune](https://cloud.githubusercontent.com/assets/576258/9124749/ce99d42c-3c91-11e5-88ba-8e443aca4634.png)

Compared to the current master:
![subsup_test_mpl_master](https://cloud.githubusercontent.com/assets/576258/9124758/dcdb5524-3c91-11e5-8173-5263f6d0f382.png)

However, not all is rosy, because the same parameters that work for CM fail miserably when using stixsans or stix as mathtext.fontset, so I will try to do a dict that sets them based on the fontset. See for stix (note the superscripts in particular):
![subsup_test_mpl_finetune-stix](https://cloud.githubusercontent.com/assets/576258/9124793/33f17960-3c92-11e5-8f4b-e3f71d81c3a9.png)

tacaswell commented:
@zblz Thank you for taking this on!

zblz commented:
So I think I've got it under control now for CM, stix and stixsans fontsets. There were a few challenges with integrals (also related to the ``Char`` vs ``Hlist`` issue above), but they look good now and I have added a couple of tests for them. I will amend the baseline images shortly and then the PR can be properly reviewed.

This is how a sample looks with this PR in cm:
![subsup_test_mpl_cm_finetune](https://cloud.githubusercontent.com/assets/576258/9133891/307d560e-3cf7-11e5-8648-dd7aa1bfbdaf.png)

To be compared with the ``usetex=True`` rendering:
![subsup_test_mpl_cm_tex](https://cloud.githubusercontent.com/assets/576258/9133729/cfeabcb0-3cf5-11e5-8474-d4942b3e37eb.png)

The same sample also looks good for stix and stixsans:
![subsup_test_mpl_stix_finetune](https://cloud.githubusercontent.com/assets/576258/9133739/dd9433be-3cf5-11e5-8bf0-f3e811dc08dd.png)
![subsup_test_mpl_stixsans_finetune](https://cloud.githubusercontent.com/assets/576258/9133740/df71ee6a-3cf5-11e5-9bfd-f3080d27a2ed.png)

For reference, this is the rendering with Computer Modern with the current master:
![subsup_test_mpl_cm_master](https://cloud.githubusercontent.com/assets/576258/9133769/25c1b558-3cf6-11e5-8cd6-364cc95aff05.png)

zblz commented:
Damn, I messed up a rebase. Please stand by...

zblz commented:
@tacaswell, @mdboom : I think this is now complete, it would be great if you could review it.

Looking at the issue list, this PR will fix the issue with spacing after superscripts reported in #4333.

zblz commented:
I have also added the sub/supercript parameters for Arev Sans, which is checked for by looking at 
```
rcParams['mathtext.fontset'] == 'custom'
rcParams['mathtext.rm'] == 'sans'
rcParams['font.sans-serif'][0].lower() == 'arev sans'
```
This could be useful for users using it and if #4636 is accepted.

zblz commented:
I just realised that doing all this I messed up the math accents and they are now shifted to the left:
![mathtext_cm_26](https://cloud.githubusercontent.com/assets/576258/9139766/4e23678e-3d24-11e5-95c1-711dcf25a7b9.png)

However, the tests do not catch this! Probably the accents are too small and don't amount to a high enough RMS. I'll try to find how I messed them up and fix it.

zblz commented:
So I found that the parameters I used for the images in the last comment above resulted in some conflicts for superscripts in slanted fonts (most notably $f$, which has weird kerning in most fonts). I have improved the superscript placement for this by adding extra kerning based on the height of the last character. It is not TeX-good, but I think TeX is doing optical kerning between the nucleus and the superscripted font, and I don't think we want to go there for a small improvement.

I have only added the png baseline images so the diff is easier to read for review. I added all the baseline images that include sub/superscripts, not only the ones failing (with PR #4887 these sets should be the same, but right now they are not). I will add the pdfs and svgs later, so Travis will fail until then.

mdboom commented:
Thanks for all this work.  I've been on vacation, but am now back and will review this soon.

mdboom commented on lib/matplotlib/mathtext.py:
```diff
@@ -2726,48 +2759,93 @@ def subsuper(self, s, loc, toks):
+                    rcParams['font.sans-serif'][0].lower() == 'Arev Sans'.lower()):
+                fs = 'arevsans'
+            else:
+                fs = 'cm'
```
For clarity, can you separate this chunk out to its own private function, perhaps called `_get_fontset_name`?

mdboom commented on lib/matplotlib/mathtext.py:
```diff
@@ -2726,48 +2759,93 @@ def subsuper(self, s, loc, toks):
+        # 7.75 and a kerning of -4.0 for an advance of 3.72, and we want to put
+        # the superscript at the advance
+        last_char = nucleus
+        if isinstance(nucleus,Hlist):
```
Space after comma

mdboom commented on lib/matplotlib/mathtext.py:
```diff
@@ -2726,48 +2759,93 @@ def subsuper(self, s, loc, toks):
-        shift_up = nucleus.height - SUBDROP * xHeight
-        if self.is_dropsub(nucleus):
-            shift_down = nucleus.depth + SUBDROP * xHeight
+        # We remove kerning for consistency (otherwise it will compute kerning
```
Maybe clarfiy: "We remove the kerning on the last character for consistency" (this doesn't remove the inter-character kerning).

mdboom commented:
This looks great to me (other than my minor comments, and the need to resolve whatever is going on with the test failures).

zblz commented:
@mdboom: I implemented your comments, and added all the rest of the baseline images (I didn't want to clutter it with the svgs and pdfs for review initially). I have replaced all baseline images that contain a subsuper, independently of whether they were failing, because current tolerance is not strict enough to fail for the position a single superscript (see #4887).

zblz commented:
And for reference here, here is some sample text as rendered with 341d481 in computer modern:

![subsup_test_mpl_cm_finetune](https://cloud.githubusercontent.com/assets/576258/9442585/84c6be8c-4a74-11e5-8bd6-bc194704c031.png)

Reference rendereing with ``usetex=True``:

![subsup_test_mpl_tex_finetune](https://cloud.githubusercontent.com/assets/576258/9442600/9de05234-4a74-11e5-91dc-7860b69000e5.png)

And for stix and stixsans:

![subsup_test_mpl_stix_finetune](https://cloud.githubusercontent.com/assets/576258/9442615/b429872c-4a74-11e5-9328-8953bc3bb9f9.png)

![subsup_test_mpl_stixsans_finetune](https://cloud.githubusercontent.com/assets/576258/9442626/b9b7d28e-4a74-11e5-9054-91028a7c628d.png)

mdboom commented:
Thanks!  I'm happy to merge this once Travis passes.

WeatherGod commented:
hold on. The second to last line has some spacing issues. The "f" and "2"
and "9" (both as a superscript and as a regular numeral) are much too close.

On Mon, Aug 24, 2015 at 10:45 AM, Michael Droettboom <
notifications@github.com> wrote:

> Thanks! I'm happy to merge this once Travis passes.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/matplotlib/matplotlib/pull/4873#issuecomment-134230522>
> .
>

WeatherGod commented:
Also, the "prime" symbol in that line is much too close, especially when
the "f" is italic.

On Mon, Aug 24, 2015 at 10:49 AM, Benjamin Root <ben.v.root@gmail.com>
wrote:

> hold on. The second to last line has some spacing issues. The "f" and "2"
> and "9" (both as a superscript and as a regular numeral) are much too close.
>
> On Mon, Aug 24, 2015 at 10:45 AM, Michael Droettboom <
> notifications@github.com> wrote:
>
>> Thanks! I'm happy to merge this once Travis passes.
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/matplotlib/matplotlib/pull/4873#issuecomment-134230522>
>> .
>>
>
>

WeatherGod commented:
Also, our mathtext might be parsing something slightly differently than
LaTeX is. In that same line, LaTeX sees the prime symbol as being on its
own and positions it as if it was a normal piece of text (nice and big),
but everywhere else, it is some sort of weird, scraggly looking superscript.

On Mon, Aug 24, 2015 at 10:49 AM, Benjamin Root <ben.v.root@gmail.com>
wrote:

> Also, the "prime" symbol in that line is much too close, especially when
> the "f" is italic.
>
> On Mon, Aug 24, 2015 at 10:49 AM, Benjamin Root <ben.v.root@gmail.com>
> wrote:
>
>> hold on. The second to last line has some spacing issues. The "f" and "2"
>> and "9" (both as a superscript and as a regular numeral) are much too close.
>>
>> On Mon, Aug 24, 2015 at 10:45 AM, Michael Droettboom <
>> notifications@github.com> wrote:
>>
>>> Thanks! I'm happy to merge this once Travis passes.
>>>
>>> —
>>> Reply to this email directly or view it on GitHub
>>> <https://github.com/matplotlib/matplotlib/pull/4873#issuecomment-134230522>
>>> .
>>>
>>
>>
>

zblz commented:
@WeatherGod: Yes, the kerning for the f is particularly tight, but it is tight everywhere, as you mention for the regular f and numerals 2 and 9 (and that's why I added them to that line), so it is not an issue with the sub/super script. I tried to minimize it for cm, where it looks pretty good, but did not manage for stix of stixsans. Maybe the kerning for the f should be slightly widened elsewhere to match TeX.

The prime issue only happens for stix and stixsans, where it seems to match $\prime$ to an already superscripted prime, as opposed to a nice and big prime for TeX and cm. This I also believe is outside the scope of this PR, but should definitely be fixed so they are consistent.

zblz commented:
I fixed a test failure in ``test_axes`` that came from having no nucleus in mathtext (e.g., 'a$^2$'), so I added a test in mathtext so it can be caught by running only the ``test_mathtext`` tests. There were also a couple of superscripts in other tests that were failing because I had not replaced the baseline images. Should be fine now.

mdboom commented:
I agree that the stix issues with the prime symbol and some of the other miscellaneous spacing is a  separate issue from this PR, and indeed this PR makes things no worse than it was before in that regard.

WeatherGod commented:
Ok, well, let's create some issue tickets for those so that we don't forget to fix them. This is great work @zblz!

WeatherGod commented:
Ok, I created issue #4987 to track the issues I noted above. Great work!

Mike, I have not reviewed a single line of code here (I wouldn't know what any of it means anyway). So, I'll leave merging this up to you?
